lightningd/htlc_end.c

```diff
@@ -1,39 +1,156 @@
 #include <ccan/cast/cast.h>
 #include <ccan/crypto/siphash24/siphash24.h>
+#include <ccan/tal/str/str.h>
 #include <ccan/tal/tal.h>
+#include <daemon/htlc.h>
+#include <daemon/log.h>
 #include <daemon/pseudorand.h>
 #include <lightningd/htlc_end.h>
+#include <stdio.h>
 
-size_t hash_htlc_end(const struct htlc_end *e)
+size_t hash_htlc_key(const struct htlc_key *k)
 {
 	struct siphash24_ctx ctx;
 	siphash24_init(&ctx, siphash_seed());
 	/* peer doesn't move while in this hash, so we just hash pointer. */
-	siphash24_update(&ctx, &e->peer, sizeof(e->peer));
-	siphash24_u64(&ctx, e->htlc_id);
-	siphash24_u8(&ctx, e->which_end);
+	siphash24_update(&ctx, &k->peer, sizeof(k->peer));
+	siphash24_u64(&ctx, k->id);
 
 	return siphash24_done(&ctx);
 }
 
-struct htlc_end *find_htlc_end(const struct htlc_end_map *map,
+struct htlc_in *find_htlc_in(const struct htlc_in_map *map,
 			       const struct peer *peer,
-			       u64 htlc_id,
-			       enum htlc_end_type which_end)
+			       u64 htlc_id)
 {
-	const struct htlc_end key = { which_end, (struct peer *)peer, htlc_id,
-				      0, NULL, NULL };
+	const struct htlc_key key = { (struct peer *)peer, htlc_id };
+	return htlc_in_map_get(map, &key);
+}
+
+static void remove_htlc_in(struct htlc_in *hend, struct htlc_in_map *map)
+{
+	htlc_in_map_del(map, hend);
+}
+
+void connect_htlc_in(struct htlc_in_map *map, struct htlc_in *hend)
+{
+	tal_add_destructor2(hend, remove_htlc_in, map);
+	htlc_in_map_add(map, hend);
+}
+
+struct htlc_out *find_htlc_out(const struct htlc_out_map *map,
+			       const struct peer *peer,
+			       u64 htlc_id)
+{
+	const struct htlc_key key = { (struct peer *)peer, htlc_id };
+	return htlc_out_map_get(map, &key);
+}
+
+static void remove_htlc_out(struct htlc_out *hend, struct htlc_out_map *map)
+{
+	htlc_out_map_del(map, hend);
+}
+
+void connect_htlc_out(struct htlc_out_map *map, struct htlc_out *hend)
+{
+	tal_add_destructor2(hend, remove_htlc_out, map);
+	htlc_out_map_add(map, hend);
+}
 
-	return htlc_end_map_get(map, &key);
+static void *PRINTF_FMT(3,4)
+	corrupt(const void *ptr, const char *abortstr, const char *fmt, ...)
+{
+	if (abortstr) {
+		char *p;
+		va_list ap;
+
+		va_start(ap, fmt);
+		p = tal_vfmt(NULL, fmt, ap);
+		fatal("%s:%s\n", abortstr, p);
```

Comment:
No, this is the main daemon (I look forward to deleting the old daemon so we can make the directory structure more sensible).

I think if we get corruption in the main daemon, we *should* restart (aborting is just a temporary hack).  Everything should be in the database (which we should check thoroughly on restart!).  If a peer is corrupt on reload, we should fire off something to salvage that peer (eg. a python script which closes onchain).